module/zfs/zfs_vnops.c

```diff
@@ -4624,32 +4622,56 @@ zfs_dirty_inode(struct inode *ip, int flags)
 
 #ifdef I_DIRTY_TIME
 	/*
-	 * This is the lazytime semantic indroduced in Linux 4.0
-	 * This flag will only be called from update_time when lazytime is set.
-	 * (Note, I_DIRTY_SYNC will also set if not lazytime)
-	 * Fortunately mtime and ctime are managed within ZFS itself, so we
-	 * only need to dirty atime.
+	 * This is the SB_LAZYTIME semantic introduced in Linux 4.0 kernel.
+	 * When the I_DIRTY_TIME flag is set and neither I_DIRTY_DATASYNC or
+	 * I_DIRTY_SYNC is set then the inode times may be written lazily.
+	 * Fortunately mtime and ctime are managed within ZFS itself, so
+	 * we only need to dirty atime.
 	 */
-	if (flags == I_DIRTY_TIME) {
+	if ((flags & I_DIRTY_TIME) &&
+	    !(flags & (I_DIRTY_SYNC | I_DIRTY_DATASYNC))) {
 		zp->z_atime_dirty = 1;
-		goto out;
+		ZFS_EXIT(zfsvfs);
+		return (0);
 	}
 #endif
 
+top:
 	tx = dmu_tx_create(zfsvfs->z_os);
 
 	dmu_tx_hold_sa(tx, zp->z_sa_hdl, B_FALSE);
 	zfs_sa_upgrade_txholds(tx, zp);
 
-	error = dmu_tx_assign(tx, TXG_WAIT);
+	/*
+	 * When a page is faulted in by filemap_page_mkwrite() this function
+	 * may be called by update_time() with the file's mmap_sem held.
+	 * Therefore it's necessary to use TXG_NOWAIT since we cannot release
+	 * the mmap_sem, and even if we could, it would be undesirable to
+	 * delay the page fault.  TXG_NOTHROTTLE will be set as needed to
+	 * bypass the write throttle.  In the unlikely case the transaction
+	 * cannot be assigned set z_atime_dirty=1 so at least the times will
+	 * be updated when the file is closed.
+	 */
+	error = dmu_tx_assign(tx, (waited ? TXG_NOTHROTTLE : 0) | TXG_NOWAIT);
```

Comment:
In `dmu_tx_try_assign`:
```
			if (dn->dn_assigned_txg == tx->tx_txg - 1) {
				mutex_exit(&dn->dn_mtx);
				tx->tx_needassign_txh = txh;
				DMU_TX_STAT_BUMP(dmu_tx_group);
				return (SET_ERROR(ERESTART));
			}
```
It doesn't seem that TXG_NOTHROTTLE will prevent dmu_tx_assign from failing? Perhaps we should bail out on `waited && error == ERESTART`?

Reply:
Good idea.  You're right, it won't prevent it from failing under all circumstances, only due to the write throttle.  I assume your specific concern is that we could end up spinning here, which would be just as bad.